### ISSUE TITLE
CA1031: Should not fire if other exceptions already caught

### ISSUE THREAD

**Issue by lonix1** [dotnet/roslyn-analyzers#6034]
### Analyzer

**Diagnostic ID**: [CA1031](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA1031)
"CA1031: Do not catch general exception types"

### Describe the improvement
````cs
try {
}
catch (FooException fe) {
}
catch (Exception e) {                  // CA1031 here
}
````

### Describe suggestions on how to achieve the rule
If one or more exceptions are already caught before the the catch-all block, the rule should not fire.

### Additional context
- .net6, vscode

**Comment by Evangelink** [comment]
Do you mind giving a little more context as to why you think the rule should not trigger in this case?

IMO (with no extra context) the rule is doing its job in reporting a catch "all" case.

**Comment by lonix1** [comment]
If the rule **always** fires, then it actually means this: "**never** have a catch-all block".

But if I already have other catch blocks, and I need that catch-all *as well*, in certain circumstances, then I've "listened" to the rule, but I still need a catch-all block in that case.

Put another way, if the rule always fires, then it implies the catch-all block should never have been part of the language. (Which is obviously wrong.)

**Comment by Evangelink** [comment]
> If the rule always fires, then it actually means this: "never have a catch-all block".

Well, I tend to think the number of catch all in an app should be close to zero so I usually add a pragma for the cases I want to keep.

> But if I already have other catch blocks, and I need that catch-all as well, in certain circumstances, then I've "listened" to the rule, but I still need a catch-all block in that case.

I see your point but I tend to think that having handled some specific cases before doesn't make the rule less relevant in this context (and so would need to be automatically suppressed). This being said, the documentation is missing some context and reasoning so that's hard to tell what's the intended scope.

@mavasani @buyaa-n Do you have some insight? Or do you have the contact of the original authors?

**Comment by mavasani** [comment]
I agree with @Evangelink here. This rule is designed to fire for every catch all block. The language supports it for rare cases where developers have no alternatives (for example you are invoking an API into another library and have no control or knowledge of what exceptions the API might throw). Catching other exception types in the same try block does not mean it is fine to catch rest of the exceptions with a catch all. You should consider suppressing the rule with a pragma suppression for cases where you feel catch all is fine in your code.
